### Task 4.8 

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\Jay\Documents\Instacart Basket Analysis'

In [3]:
# import merged data
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data Sets', 'Prepared Data', 'ords_prods_merge_new.pkl'))

In [4]:
ords_prods_merge.shape

(32435059, 20)

In [5]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2539329        1    prior             1                   2   
2   2539329        1    prior             1                   2   
3   2539329        1    prior             1                   2   
4   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_prior_order  new_customer  product_id  \
0                  8                     NaN          True         196   
1                  8                     NaN          True       14084   
2                  8                     NaN          True       12427   
3                  8                     NaN          True       26088   
4                  8                     NaN          True       26405   

   add_to_cart_order  reordered                             product_name  \
0                  1          0                                     Soda   
1                  2          0  Organic Unsweetened Vanilla Almond Milk   
2                  3          0                      Original Beef Jerky   
3                  4          0               Aged White Cheddar Popcorn   
4                  5          0         XL Pick-A-Size Paper Towel Rolls   

   aisle_id  department_id  prices _merge    price_range_loc     Busiest day  \
0      77.0            7.0     9.0   both  Mid-range product  Regularly busy   
1      91.0           16.0    12.5   both  Mid-range product  Regularly busy   
2      23.0           19.0     4.4   both  Low-range product  Regularly busy   
3      23.0           19.0     4.7   both  Low-range product  Regularly busy   
4      54.0           17.0     1.0   both  Low-range product  Regularly busy   

     Busiest_days busiest_period_of_day  
0  Regularly busy           Most_orders  
1  Regularly busy           Most_orders  
2  Regularly busy           Most_orders  
3  Regularly busy           Most_orders  
4  Regularly busy           Most_orders

In [6]:
# new grouping function
ords_prods_merge.groupby('product_name')

In [7]:
# single aggregation
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1.0              15.457838
2.0              17.277920
3.0              17.170395
4.0              17.811403
5.0              15.215751
6.0              16.439806
7.0              17.225802
8.0              15.340650
9.0              15.895474
10.0             20.197148
11.0             16.170638
12.0             15.887671
13.0             16.583536
14.0             16.773669
15.0             16.165037
16.0             17.665606
17.0             15.694469
18.0             19.310397
19.0             17.177343
20.0             16.473447
21.0             22.902379

##### The mean values are comparable and only differ slightly, usually a bit higher, or lower in a few cases, by compaison to those of the subset shown in the exercise. 

In [8]:
# aggregating w/transform()
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [9]:
# remove display limit
pd.options.display.max_rows = None

In [10]:
ords_prods_merge.head(100)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2539329        1    prior             1                   2   
2    2539329        1    prior             1                   2   
3    2539329        1    prior             1                   2   
4    2539329        1    prior             1                   2   
5    2398795        1    prior             2                   3   
6    2398795        1    prior             2                   3   
7    2398795        1    prior             2                   3   
8    2398795        1    prior             2                   3   
9    2398795        1    prior             2                   3   
10   2398795        1    prior             2                   3   
11    473747        1    prior             3                   3   
12    473747        1    prior             3                   3   
13    473747        1    prior             3                   3   
14    473747        1    prior             3                   3   
15    473747        1    prior             3                   3   
16   2254736        1    prior             4                   4   
17   2254736        1    prior             4                   4   
18   2254736        1    prior             4                   4   
19   2254736        1    prior             4                   4   
20   2254736        1    prior             4                   4   
21    431534        1    prior             5                   4   
22    431534        1    prior             5                   4   
23    431534        1    prior             5                   4   
24    431534        1    prior             5                   4   
25    431534        1    prior             5                   4   
26    431534        1    prior             5                   4   
27    431534        1    prior             5                   4   
28    431534        1    prior             5                   4   
29   3367565        1    prior             6                   2   
30   3367565        1    prior             6                   2   
31   3367565        1    prior             6                   2   
32   3367565        1    prior             6                   2   
33    550135        1    prior             7                   1   
34    550135        1    prior             7                   1   
35    550135        1    prior             7                   1   
36    550135        1    prior             7                   1   
37    550135        1    prior             7                   1   
38   3108588        1    prior             8                   1   
39   3108588        1    prior             8                   1   
40   3108588        1    prior             8                   1   
41   3108588        1    prior             8                   1   
42   3108588        1    prior             8                   1   
43   3108588        1    prior             8                   1   
44   2295261        1    prior             9                   1   
45   2295261        1    prior             9                   1   
46   2295261        1    prior             9                   1   
47   2295261        1    prior             9                   1   
48   2295261        1    prior             9                   1   
49   2295261        1    prior             9                   1   
50   2550362        1    prior            10                   4   
51   2550362        1    prior            10                   4   
52   2550362        1    prior            10                   4   
53   2550362        1    prior            10                   4   
54   2550362        1    prior            10                   4   
55   2550362        1    prior            10                   4   
56   2550362        1    prior            10                   4   
57   2550362        1    prior            10                

In [11]:
# set flags w/loc()
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [12]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [13]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [14]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891507
Loyal customer      10294027
New customer         6249525
Name: loyalty_flag, dtype: int64

In [15]:
# loyalty flag basic stats
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

##### The stats show new customers initially spending the most at 13.29, long term loyal customer average spending droping to around 10.39, with regular customer average spending slightly less than the newbies at 12.50. The max value 99999 still needs investigation, as the exercise says InstaCart limits the top price to 99.

In [16]:
# create user spending flag 
ords_prods_merge['spending'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [17]:
ords_prods_merge.loc[ords_prods_merge['spending'] < 10, 'spending_flag'] = 'Low Spender'

In [18]:
ords_prods_merge.loc[ords_prods_merge['spending'] >= 10, 'spending_flag'] = 'High Spender'

In [19]:
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low Spender     31800381
High Spender      634678
Name: spending_flag, dtype: int64

In [24]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2539329        1    prior             1                   2   
2   2539329        1    prior             1                   2   
3   2539329        1    prior             1                   2   
4   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_prior_order  new_customer  product_id  \
0                  8                     NaN          True         196   
1                  8                     NaN          True       14084   
2                  8                     NaN          True       12427   
3                  8                     NaN          True       26088   
4                  8                     NaN          True       26405   

   add_to_cart_order  ...  prices _merge    price_range_loc     Busiest day  \
0                  1  ...     9.0   both  Mid-range product  Regularly busy   
1                  2  ...    12.5   both  Mid-range product  Regularly busy   
2                  3  ...     4.4   both  Low-range product  Regularly busy   
3                  4  ...     4.7   both  Low-range product  Regularly busy   
4                  5  ...     1.0   both  Low-range product  Regularly busy   

     Busiest_days busiest_period_of_day max_order  loyalty_flag  spending  \
0  Regularly busy           Most_orders        10  New customer  6.367797   
1  Regularly busy           Most_orders        10  New customer  6.367797   
2  Regularly busy           Most_orders        10  New customer  6.367797   
3  Regularly busy           Most_orders        10  New customer  6.367797   
4  Regularly busy           Most_orders        10  New customer  6.367797   

  spending_flag  
0   Low Spender  
1   Low Spender  
2   Low Spender  
3   Low Spender  
4   Low Spender  

[5 rows x 24 columns]

In [25]:
# create order frequency flag
ords_prods_merge['frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [26]:
ords_prods_merge.loc[ords_prods_merge['frequency'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [27]:
ords_prods_merge.loc[(ords_prods_merge['frequency'] > 10) & (ords_prods_merge['frequency'] <= 20), 'frequency_flag'] = 'Regular customer'

In [28]:
ords_prods_merge.loc[ords_prods_merge['frequency'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [34]:
ords_prods_merge[['user_id','frequency','frequency_flag']].head(100)

user_id  frequency         frequency_flag
0         1       20.5  Non-frequent customer
1         1       20.5  Non-frequent customer
2         1       20.5  Non-frequent customer
3         1       20.5  Non-frequent customer
4         1       20.5  Non-frequent customer
5         1       20.5  Non-frequent customer
6         1       20.5  Non-frequent customer
7         1       20.5  Non-frequent customer
8         1       20.5  Non-frequent customer
9         1       20.5  Non-frequent customer
10        1       20.5  Non-frequent customer
11        1       20.5  Non-frequent customer
12        1       20.5  Non-frequent customer
13        1       20.5  Non-frequent customer
14        1       20.5  Non-frequent customer
15        1       20.5  Non-frequent customer
16        1       20.5  Non-frequent customer
17        1       20.5  Non-frequent customer
18        1       20.5  Non-frequent customer
19        1       20.5  Non-frequent customer
20        1       20.5  Non-frequent customer
21        1       20.5  Non-frequent customer
22        1       20.5  Non-frequent customer
23        1       20.5  Non-frequent customer
24        1       20.5  Non-frequent customer
25        1       20.5  Non-frequent customer
26        1       20.5  Non-frequent customer
27        1       20.5  Non-frequent customer
28        1       20.5  Non-frequent customer
29        1       20.5  Non-frequent customer
30        1       20.5  Non-frequent customer
31        1       20.5  Non-frequent customer
32        1       20.5  Non-frequent customer
33        1       20.5  Non-frequent customer
34        1       20.5  Non-frequent customer
35        1       20.5  Non-frequent customer
36        1       20.5  Non-frequent customer
37        1       20.5  Non-frequent customer
38        1       20.5  Non-frequent customer
39        1       20.5  Non-frequent customer
40        1       20.5  Non-frequent customer
41        1       20.5  Non-frequent customer
42        1       20.5  Non-frequent customer
43        1       20.5  Non-frequent customer
44        1       20.5  Non-frequent customer
45        1       20.5  Non-frequent customer
46        1       20.5  Non-frequent customer
47        1       20.5  Non-frequent customer
48        1       20.5  Non-frequent customer
49        1       20.5  Non-frequent customer
50        1       20.5  Non-frequent customer
51        1       20.5  Non-frequent customer
52        1       20.5  Non-frequent customer
53        1       20.5  Non-frequent customer
54        1       20.5  Non-frequent customer
55        1       20.5  Non-frequent customer
56        1       20.5  Non-frequent customer
57        1       20.5  Non-frequent customer
58        1       20.5  Non-frequent customer
59        2       13.0       Regular customer
60        2       13.0       Regular customer
61        2       13.0       Regular customer
62        2       13.0       Regular customer
63        2       13.0       Regular customer
64        2       13.0       Regular customer
65        2       13.0       Regular customer
66        2       13.0       Regular customer
67        2       13.0       Regular customer
68        2       13.0       Regular customer
69        2       13.0       Regular customer
70        2       13.0       Regular customer
71        2       13.0       Regular customer
72        2       13.0       Regular customer
73        2       13.0       Regular customer
74        2       13.0       Regular customer
75        2       13.0       Regular customer
76        2       13.0       Regular customer
77        2       13.0       Regular customer
78        2       13.0       Regular customer
79        2       13.0       Regular customer
80        2       13.0       Regular customer
81        2       13.0       Regular customer
82        2       13.0       Regular customer
83        2       13.0       Regular customer
84        2       13.0       Regular customer
85        2       13.0       Regular customer
86

In [35]:
ords_prods_merge.shape

(32435059, 26)

##### Not certain why there are apparently multiple identical records for user_id 1, 2, etc. Needs investigation. 

In [37]:
# export to pkl
ords_prods_merge.to_pickle(os.path.join(path, '02 Data Sets','Prepared Data', 'ords_prods_merge_agg.pkl'))